debugger/src/debug_adapter.rs

```diff
@@ -196,25 +224,321 @@ impl<P: ProtocolAdapter> DebugAdapter<P> {
+                        bytes_written: Some(data_bytes.len() as i64),
+                        offset: None,
+                    })),
+                )?;
+                // TODO: This doesn't trigger the UI to reload the variables effected. Investigate if we can force it in some other way, or if it is a known issue.
+                self.send_event(
+                    "memory",
+                    Some(MemoryEventBody {
+                        count: data_bytes.len() as i64,
+                        memory_reference: format!("{:#010x}", address),
+                        offset: 0,
+                    }),
+                )
+            }
             Err(error) => self.send_response::<()>(request, Err(error)),
         }
     }
+
+    /// Evaluates the given expression in the context of the top most stack frame.
+    /// The expression has access to any variables and arguments that are in scope.
+    /// TODO: When variables appear in the `watch` context, they will not resolve correctly after a 'step' function. Consider doing the lazy load for 'either/or' of Variables vs. Evaluate
+    pub(crate) fn evaluate(&mut self, core_data: &mut CoreData, request: Request) -> Result<()> {
+        let arguments: EvaluateArguments = match self.adapter_type() {
+            DebugAdapterType::CommandLine => todo!(),
+            // match request.arguments.as_ref().unwrap().try_into() {
+            //     Ok(arguments) => arguments,
+            //     Err(error) => return self.send_response::<()>(request, Err(error)),
+            // },
+            DebugAdapterType::DapClient => match get_arguments(&request) {
+                Ok(arguments) => arguments,
+                Err(error) => return self.send_response::<()>(request, Err(error)),
+            },
+        };
+
+        // Various fields in the response_body will be updated before we return.
+        let mut response_body = EvaluateResponseBody {
+            indexed_variables: None,
+            memory_reference: None,
+            named_variables: None,
+            presentation_hint: None,
+            result: format!("<variable not found {:?}>", arguments.expression),
+            type_: None,
+            variables_reference: 0_i64,
+        };
+
+        // The Variables request always returns a 'evaluate_name' = 'name', this means that the expression will always be the variable name we are looking for.
+        let expression = arguments.expression.clone();
+
+        // Make sure we have a valid StackFrame
+        if let Some(stack_frame) = match arguments.frame_id {
+            Some(frame_id) => core_data
+                .stack_frames
+                .iter_mut()
+                .find(|stack_frame| stack_frame.id == frame_id),
+            None => {
+                // Use the current frame_id
+                core_data.stack_frames.first_mut()
+            }
+        } {
+            // Always search the registers first, because we don't have a [VariableCache] for them.
```

Comment:
```suggestion
            // Always search the registers first, because we don't have a VariableCache for them.
```

Crosslinks do not do anything in normal comments :)